src/prebuild/main.ml

```diff
@@ -0,0 +1,208 @@
+	(* params *) get_optional_field "params" as_params [] fields,
+	(* platforms *) get_optional_field "platforms" as_platforms [] fields,
+	(* targets *) get_optional_field "targets" as_targets [] fields,
+	(* internal *) get_optional_field "internal" as_bool false fields
+
+let parse_file_array path map =
+	let file = open_in path in
+	let data = Std.input_all file in
+	let open Json.Reader in
+	let lexbuf = Sedlexing.Utf8.from_string data in
+	let json = read_json lexbuf in
+	match json with
+	| JArray s -> List.map map s
+	| _ -> raise (Prebuild_error "not an array")
+
+let gen_platforms = function
+	| [] -> []
+	| [platform] -> ["Platform " ^ platform]
+	| platforms -> ["Platforms [" ^ (String.concat ";" platforms) ^ "]"]
+
+let gen_params = List.map (function param -> "HasParam \"" ^ param ^ "\"" )
+
+let gen_define_type defines =
+	String.concat "\n" (List.map (function (name, _, _, _, _) -> "\t| " ^ name) defines)
+
+let gen_define_info defines =
+	let define_str = List.map (function
+		(name, define, doc, params, platforms) ->
+			let platforms_str = gen_platforms platforms in
+			let params_str = gen_params params in
+			"\t| " ^ name ^ " -> \"" ^ define ^ "\",(" ^ (Printf.sprintf "%S" doc) ^ ",[" ^ (String.concat "; " (platforms_str @ params_str)) ^ "])"
+	) defines in
+	String.concat "\n" define_str
+
+let gen_meta_type metas =
+	String.concat "\n" (List.map (function
+		| ("InlineConstructorArgument", _, _, _, _, _, _) -> "\t| InlineConstructorArgument of int * int"
+		| (name, _, _, _, _, _, _) -> "\t| " ^ name
+	) metas)
+
+let gen_meta_info metas =
+	let meta_str = List.map (function
+		(name, metadata, doc, params, platforms, targets, internal) ->
+			let platforms_str = gen_platforms platforms in
+			let params_str = gen_params params in
+			let targets_str = (match targets with
+				| [] -> []
+				| [target] -> ["UsedOn " ^ target]
+				| targets -> ["UsedOnEither [" ^ (String.concat ";" targets) ^ "]"]
+			) in
+			let internal_str = if internal then ["UsedInternally"] else [] in
+			let name = (match name with
+				(* this is a hacky, I know *)
+				| "InlineConstructorArgument" -> "InlineConstructorArgument _"
+				| _ -> name
+			) in
+			"\t| " ^ name ^ " -> \"" ^ metadata ^ "\",(" ^ (Printf.sprintf "%S" doc) ^ ",[" ^ (String.concat "; " (platforms_str @ params_str @ targets_str @ internal_str)) ^ "])"
+	) metas in
+	String.concat "\n" meta_str
+
+let autogen_header = "(* This file is auto-generated using prebuild from files in src-json *)
+(* Do not edit manually! *)
+"
+
+let define_header = autogen_header ^ "
+open Globals
+
+type define_parameter =
+	| HasParam of string
+	| Platform of platform
+	| Platforms of platform list
+
+"
+
+let meta_header = autogen_header ^ "
+open Globals
+
+type meta_usage =
+	| TClass
+	| TClassField
+	| TAbstract
+	| TAbstractField
+	| TEnum
+	| TTypedef
+	| TAnyField
+	| TExpr
+	| TTypeParameter
+	| TVariable
+
+type meta_parameter =
+	| HasParam of string
+	| Platform of platform
```

Comment:
Is there a reason to keep both `Platform` and `Platforms` variants around? I thought that was mostly for more convenient declaration in ML sources, but now it's in JSON.

Same for "used on".

Reply:
The only difference as far as I can see is the help outputs: `Platforms [Java;Cs]` produces `(for java,cs)`, `Platform Java` produces `(java only)`.

Reply:
Right, that can just be done by checking the length of the array though.